---
title: "02_DESEQ_differential_expression"
author: "JR"
date: "8/9/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
# library(tximport)
# library(ggrepel)
library(pheatmap)
source("../../../../../util/plotting_functions.R")
source("../../../../../util/_setup.R")

# DEseq2 needs a specific package of 'locfit' that has to be installed first
install.packages("https://cran.r-project.org/src/contrib/Archive/locfit/locfit_1.5-9.4.tar.gz", repos=NULL, type="source")

# Now we can install DEseq2
BiocManager::install("DESeq2")

# now load the package
library(DESeq2)

```
# Differential Expression (nuclear versus cytoplasmic)
Above we got to normalized counts that are ready to compared across
conditions. Note we "factored" the sample sheet so each fraction has
a 'factor' value. So we can now factor nuclear and compare to cytoplasmic.

Let's specialize sample sheet we can use to compare nuc/cyto samples.

```{R nuclear and cytoplasmic samplesheet}

# Now we will filter the sample sheet to just nuclear and cytoplasmic
samplesheet_nuc_cyto <- samplesheet %>%
  filter(condition %in% c("homo_sapiens_cytosolic_fraction", "homo_sapiens_nuclear_fraction"))
# Nice now we have a reduced sample sheet that is well organized.

# Next, let's triple check that the counts data
# contains the counts for the samples in the reduced sample sheet 
samplesheet_nuc_cyto$sample_name %in% colnames(counts)
# Cool, the 4 samples in filtered are same as counts

# Again we need to think about "rownames"
# Note the rownames are numbers -- this is 'meta' information we need.
# To make rownames we will turn into data.frame and add rownames
samplesheet_nuc_cyto <- samplesheet_nuc_cyto %>%
  as.data.frame()

# IMPORTANT: adding rownames from nuc_cyto_samplesheet by indexing
rownames(samplesheet_nuc_cyto) <- samplesheet_nuc_cyto$sample_name
# Now if we look at this object we see rownames as sample_name - nice.

# NOTE: DeSeq wants the control as a factor (fct), currently (chr)
# The first factor level is the "control" and all the other levels
# will be compared back to it.

# Let's make the condition a factor:
samplesheet_nuc_cyto$condition <- factor(samplesheet_nuc_cyto$condition, 
                                         levels = c("homo_sapiens_nuclear_fraction",
                                                    "homo_sapiens_cytosolic_fraction"))


# let's write out this samplesheet for future use
write_rds(samplesheet_nuc_cyto, "samplesheet_nuc_cyto.rds")

```

# START POINT
All the work we have done above can be skipped by 
reading in the following files.

# To learn more about DeSeq and other variations not used here:
https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/08_practical.pdf

```{r START POINT}

# Read in the raw counts matrix
counts <- read_rds("counts_matrix.rds")

# Read in rlog normalized counts 
rlog_counts_matrix <- read_rds("rlog_counts_all.rds")

# full sample_sheet
samplesheet <- read_rds("final_samplesheet.rds")

# nuc/cyto sample sheet
samplesheet_nuc_cyto <- read_rds("samplesheet_nuc_cyto.rds")

# finally a gene_id and gene_names object
g2s <- read.csv("g2s.csv")

```

# Heatmap of all samples
First let's cluster all the data to see how the samples relate.
```{r heatmap of all data}

# Now we can make a heatmap of all the different fractions
# First let's look at the rlogCounts
hist(rlog_counts_matrix, breaks = 100, xlim = c(0,20))

# First we will filter to genes that are variable across data.
# rowVars gives us the variance across samples
hist(rowVars(rlog_counts_matrix), breaks = 100, xlim = c(0,4))

# let's look at this as an object
x <- rowVars(rlog_counts_matrix)


# Filterting out genes that don't have variance of more than 2
rlog_var_genes_all <- rlog_counts_matrix[rowVars(rlog_counts_matrix) > 2,]

# row center the counts -- we need to flip the matrix
# we can only scale cols so we need to transmute (t)
# then turn it back with (t)
scaled_counts <- t(scale(t(rlog_var_genes_all))) %>%
  as.matrix()

# make heat map
pheatmap(scaled_counts, show_rownames = FALSE)

# saving
pdf("figures/all_vs_total_genes_heatmap.pdf")
pheatmap(scaled_counts, show_rownames = FALSE)
dev.off()
```

# Principal component analysis
Now let's use Principle Component Analysis to determine sample similarity

```{r PCA analysis}

# The first thing that we'll want to do is make a PCA plot.
# 'prcomp()' is the principle component analysis function.

?prcomp

# let's calculate the values associated with 'prcomp'
# install.packages("ggrepel") to move text so it doesn't overlap

# First we need to flip the counts matrix as PCA needs to be across rows (samples)
rld_pca <- prcomp(t(rlog_counts_matrix))

# setting up a data frame with column x 
rld_prcomps <- rld_pca$x %>% as.data.frame() %>%
  
  # note that x is in rld_pca as the PCA values
  rownames_to_column("sample_name") %>%
  # selecting data we want to merge with samplesheet
  select(sample_name, PC1, PC2)
  # merging PC1 and PC2 with sample sheet via sample name
rld_prcomps <- merge(samplesheet, rld_prcomps)



# install.packages("ggrepel")


# Let's make a PCA plot
ggplot(rld_prcomps, aes(x = PC1, y = PC2, color = condition, label = replicate)) +
geom_point() +
geom_text_repel()
# RESULT: PCA plot looks like replicates are more similar than fraction type!

# Let's print this out
ggsave("figures/PCA_plot.pdf")

```

# Nuclear versus cytoplasmic differential expression
Next we will reduce out counts_matrix to just the values of nuc/cyto

```{r nuc/cyto  counts}

# first reduce the counts matrix to just nuc and cyto samples.
counts_nuc_cyto <- counts[, samplesheet_nuc_cyto$sample_name]

# This is a good way to triple check everything is in order.

rownames(samplesheet_nuc_cyto) == colnames(counts_nuc_cyto)
# Great this evaluated to true and thus moved on -- we would get an error message otherwise.

# or the other way we have been doing this:
rownames(samplesheet_nuc_cyto) == colnames(counts_nuc_cyto)

```

# Finalinzing samplesheet for nuc vs cyto DEseq2
We are almost ready, last things to check:

1) DeSeq2 requires integer data 

2) Deseq2 needs the condition being compared to as a "factor" 
-- this is critical --

```{R organizing nuc/cyto counts and samplesheet for deseq}

# Ok now we need to turn the counts into integer data (int)

# The first step would be to round the counts, but we have already done that.
# But good to remember to round before setting chr to int.
# counts_nuc_cyto <- round(counts_nuc_cyto)

# mode can convert chr to int, let's change everything from chr to int
mode(counts_nuc_cyto) <- "integer"

# Let's fileter to genes that have at least one count across cols.
nuc_cyto_counts_filtered <- counts_nuc_cyto[rowSums(counts_nuc_cyto) > 1,]

```

# Differential expression of nuc -vs- cyto DEseq2
Now let's look deeper into the results held in dds object

```{r DEseq2 nuc versus cyto}

# IMPORTANT: always double check organization of input files to DESeq!
# If this is not true the results will be a mix of wrong samples!
# We did this above but double checking here
colnames(nuc_cyto_counts_filtered) == rownames(samplesheet_nuc_cyto)

# looks good, now let's make the required DEseq raw counts
dds <- DESeqDataSetFromMatrix(countData = nuc_cyto_counts_filtered,
                              colData = samplesheet_nuc_cyto,
                              design = ~ condition)


dds <- DESeq(dds)

# IMPORTANT : we need the results name for exporting results to data.frame
# We can see that DeSeq makes it's own names so "resultNames" is the way to get that info
resultsNames(dds)

# We don't want the intercept data
# but we need to extract the results from dds:
# let's remove intercept:

res <- results(dds, name = "condition_homo_sapiens_cytosolic_fraction_vs_homo_sapiens_nuclear_fraction")

# It's easier to view it as a data.frame so we'll convert it.
# While we are at it we merge in gene_symbols (g2s)

res_df <- res %>% as.data.frame() %>%
  # moving the row values to a "meta" column
  rownames_to_column("gene_id") %>%
  # merging in gene_name with gene_id
  merge(g2s)

# Nice now we have a DF with a lot of good info for each gene.
# Let's write it out.

write_rds(res_df, "results/nuclear_cyto_results_df.rds")


```

QC plot: Volcano 
```{r valono, MA and other plots}

# Test is CYTO/Nuclear
# so larger values are cyto and smaller nuclear

# A first check is a "volcano" plot:
# x-axis = log fold change
# y-axis = pval

# Plotting volcano plot of nuclear vs cytoplasmic

ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj), color = -log10(res_df$padj) > 90)) + 
  geom_point()

# saving volcano plot
ggsave("figures/nuc_cyto_volcano.pdf")


# Note there was a gene log2 -30 fold down meaning very nuclear
nuclear_gene <- res_df$gene_name[which.min(res_df$log2FoldChange)]
# Nice U2 is very nuclear and good control

# we can find most cytoplasmic here too
cyto_gene <- res_df$gene_name[which.max(res_df$log2FoldChange)]
# 7SK -- a lncRNA ...


# Making a list of nuclear genes at certain threshold
nuclear_genes <- res_df %>% 
  filter(log2FoldChange < -1, padj < 0.05) %>%
  as.data.frame()

write_csv(nuclear_genes, "nuclear_genes.csv")

# Cytoplasmic genes:
cyto_genes <- res_df %>%
  filter(log2FoldChange > 1, padj < 0.05)

write_csv(cyto_genes, "cyto_genes.csv")


```

# MA-PLOTS are one of the common first plots to make.
# This plot informs you of low abundance biases in fold-change

```{R MA PLOT}

# The 'MA plot' shows how low abundance affects can be an issue.
# x-axis = log(expression level)
# y-axis = log fold change

ggplot(res_df, aes(x = log10(baseMean), y = log2FoldChange)) + 
  geom_point()

# This is a good QC plot and shows the fold changes
# are not biased to low "base-mean" levels

# saving MA PLOT

ggsave("figures/MA_plot_nuc_cyto.pdf")


# Now let's look at nuclear MA plot seperate:
ggplot(nuclear_genes, aes(x = baseMean , y = log2FoldChange)) + 
  xlim(0, 50000) +
  geom_point()

```

let's see the range of pvalues in nuc-vs-cyto:
```{R pvalues nuc-vs-cyto / FDR example}

# let's look at the distribution of pvalues

hist(res_df$padj)

# we could apply a false discovery as we see a tail of distributed pvalues
```

# DESeq with all subcellular fractions

```{r organizing sample sheet for comparison to total RNA for each fraction}

# We will use total RNA (whole cell) condition to compare everything back to.

# we will want to set the factor levels with "total" first
samplesheet$condition <- factor(samplesheet$condition, levels = c("homo_sapiens_hepg2", "homo_sapiens_membrane_fraction", "homo_sapiens_insoluble_cytoplasmic_fraction", "homo_sapiens_cytosolic_fraction", "homo_sapiens_nuclear_fraction"))


# creating dds based on the factors above
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = samplesheet,
                              design = ~ condition)


# Run the DESeq stats
dds <- DESeq(dds)

```

# What is in DDS
Let's look a little deeper into the dds results from total vs all
```{r looking at results from dds }

# We now have a bunch more results embedded in the dds object
resultsNames(dds)

# let's take this name to return one set of the results from DeSeq
# all comparisions are in our dds so we can select one at a time.

# Let's just look at one of the results
res <- results(dds, name = "condition_homo_sapiens_membrane_fraction_vs_homo_sapiens_hepg2")

# We can further index the information for just this sample:
res_df <- res %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s) %>%
  mutate(result_name = "condition_membrane_fraction_vs_total")

# Looking to see log2 fold change range
summary(res_df$log2FoldChange)

```


Ok that was a lot of RNAseq !!
Let's save the results:
```{r compiling all the results}

# A good starting point would be to use a for loop to make a data.frame 
# with all the results
results_names <- resultsNames(dds)

# let's look
results_names

# We don't care about the intercept, so we can leave that out
results_names <- results_names[-1]


#TODO ? How does the for loop fill this data frame? 
# First we set up a data frame for logFC results.
res_df <- data.frame("gene_id" = character(), 
                     "baseMean" = numeric(), 
                     "log2FoldChange" = numeric(), 
                     "lfcSE" = numeric(),
                     "stat" = numeric(),
                     "pvalue" = numeric(),
                     "padj" = numeric(),
                     "gene_name" = character(),
                     "result_name" = character())

# in parallel we can make the same results with "shrunken"
# logFC this normalizes low expressed genes to be less significant.

res_shrunken_df <- data.frame("gene_id" = character(), 
                              "baseMean" = numeric(), 
                              "log2FoldChange" = numeric(), 
                              "lfcSE" = numeric(),
                              "stat" = numeric(),
                              "pvalue" = numeric(),
                              "padj" = numeric(),
                              "gene_name" = character(),
                              "result_name" = character())

# Now we will make a forloop to populate these data.frames with each DeSeq result !
# We will also apply a shrunken log fold change 

for(i in 1:length(results_names)) {
  # grabbing the name of the result file i
  results_name <- results_names[i]
  # populating the res_df with results(dds)
  # x <- results(dds, name = results_name)
  res <- results(dds, name = results_name)
  # populating res shrunken lfc with flcShrink
  # Y <- lfcShrink(dds, coef = results_name,  res = res)
  res_shrunken <- lfcShrink(dds, coef = results_name,  res = res)
  
  # populating data.frame 1 : temp_res_df
  tmp_res_df <- res %>% as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    merge(g2s) %>%
    mutate(result_name = results_name)
  
  # populating data.frame 1 : temp_res_shrunken
  tmp_res_shrunken_df <- res_shrunken %>% as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    merge(g2s) %>%
    mutate(result_name = results_name)
  
  # Append to full data.frame
  res_df <- bind_rows(res_df, tmp_res_df)
  res_shrunken_df <- bind_rows(res_shrunken_df, tmp_res_shrunken_df)
}

# Let's save these res_df
write_rds(res_df, "results/deseq_results_df.rds")

# shrunken log fold change results
write_rds(res_shrunken_df, "results/deseq_results_shrunken_lfc_df.rds")

```

Cool now we have two great data frames to move forward with.
Normally we would just move forward with shrunken logFC.
Let's add a couple more fun colums to this res_shrunken_df

# START POINT Nuclear cytoplasmic comparison

```{r signigicant nuclear and cyto}

# reading in nuc_cyto results df
nuc_cyto_res_df <- read_rds("results/nuclear_cyto_results_df.rds")

# Let's make list of all SIG gnees with  P <0.01 and FC > 2 in nuc-vs-cyto
nuc_cyto_sig_genes <- nuc_cyto_res_df %>%
  filter(padj < 0.001, abs(log2FoldChange) >1)

# Now we filter res_shrunken for gene_id column and subcellular fraction column 
nuc_cyto_genes_to_plot <- res_shrunken_df %>%
  filter(gene_id %in% nuc_cyto_sig_genes$gene_id, result_name %in% c("condition_homo_sapiens_nuclear_fraction_vs_homo_sapiens_hepg2", "condition_homo_sapiens_cytosolic_fraction_vs_homo_sapiens_hepg2"))
  
# We need a matrix for heatmap and converting nuc_cyto_genes/values from lines above to matrix
lfc_matrix <- nuc_cyto_genes_to_plot %>% 
  dplyr::select(gene_id, log2FoldChange, result_name) %>% 
  pivot_wider(names_from = "result_name", values_from = "log2FoldChange") %>%
  column_to_rownames("gene_id") %>%
  as.matrix()

# Let's plot it with pheatmap
pheatmap::pheatmap(lfc_matrix, show_rownames = FALSE, breaks = seq(-3, 3, length.out = 100))

# saving
pdf("figures/nuc_vs_cyto_genes_heatmap.pdf")
pheatmap::pheatmap(lfc_matrix, show_rownames = FALSE, breaks = seq(-3, 3, length.out = 100))
dev.off()

```