---
title: "01_create_consensus_peaks"
author: "JR"
date: "8/2/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(GenomicRanges)
library(tidyverse)
library(IRanges)
source("../../util/my_class_functions.R")

# filepath to import peaks
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/<your_folder>"
peak_path <- "CLASS_2023/CLASSES/03_Nextflow/00_my_chipseq/results/bwa/mergedLibrary/macs/broadPeak/"
broadpeakfilepath <- file.path(basepath, peak_path)
```

Goal: to create a consensus_peak file for each dbp, export and format for UCSC.
The first step in analyses is looking at raw data :)
Here we will take our functions to create a consensus peak file and see how it 
lines up with raw data on the UCSC genome browser. We will do this in these steps:

1) Import peaks
2) create consensus peaks
3) export named GRanges list to .bed files
4) format the file to be uploaded into UCSC
5) inspect raw data to see if consensus peaks make sense.



# STEP1: import peaks 

```{r import peaks}

# import peaks
peak_list <- import_peaks(consensus_file_path = broadpeakfilepath)

# let's get a list of how many peaks are in each file before we create consensus peaks.
peak_num <- sapply(peak_list, length) %>% as.data.frame(row.names = T)
# label column
names(peak_num) <- c("num_peaks")

# make dbp name a col.

peak_num <- peak_num %>%
  rownames_to_column(var = "dbp") %>%
  separate(col = dbp,  into = c('dbp', 'replicate'), sep = "_")
  
# peak_num <- separate(peak_num, col = dbp,  into = c('dbp', 'replicate'), sep = "_")


# let's save this for our records 
write_csv(peak_num, "results/num_peaks_df.csv")

# cool our first results file !
```


# STEP2: create consensus peaks
```{r create consensus peaks}
# first we need a list of unique dbps as input parameter
dbps <- unique(sapply(names(peak_list), function(x) {
   unlist(strsplit(x, "_"))[1]
}))

# now run our function consensus_from_reduced
consensus_list <- lapply(dbps, consensus_from_reduced, peak_list)
names(consensus_list) <- dbps

```

# let's gather some data about peak numbers and save
```{r num_peaks data frame}
# nice let's see the num_peaks for consensus 
num_consensus_peaks <- sapply(consensus_list, length) %>% 
  as.data.frame() %>%
  rownames_to_column( var = "dbp") %>%
  dplyr::rename(number_consensus_peaks = ".")


# merge into num_peaks
peak_num <- left_join(peak_num, num_consensus_peaks)

# saving 
write_csv(peak_num, "results/num_peaks_df.csv")

```

#STEP 3: export consensus files
Since we will be using these quite often let's make a directory for them
Let's use "analysis/00_consensus_peaks/

```{r exporting consensus peaks }
# setting file path to export
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/<your_folder>"
consensus_path <- "CLASS_2023/CLASSES/05_R_analyses/analysis/00_consensus_peaks/"
exportpath <- file.path(basepath, consensus_path)

# now we can export each as .bed file
for(i in 1:length(consensus_list)) {
rtracklayer::export(consensus_list[[i]], paste0(exportpath, names(consensus_list)[i], "_consensus_peaks.bed") )}

# alternative local export:
for(i in 1:length(consensus_list)) {
rtracklayer::export(consensus_list[[i]], paste0("results/consensus_peaks/test", names(consensus_list)[i],"_consensus_peaks.bed")) }

```


# STEP 4 : format .bed files in general and add header needed for UCSC.
Now let's read in one of these files and see what we see:

```{r reading .bed file from rtracklayer_export}
# read in a given peak file 
peak_1 <- read_tsv("analysis/00_consensus_peaks/EP300_consensus_peaks.bed", col_names = F)

# ok we can see that rtracklayer did not keep qvalue sigvalue etc
# we also see the odd contings etc.
# let's clean this up with lapply:
```

# STEP4_2 : cleaning up .bed file basics (col anmes and chromosomes)
```{r reading .bed file from rtracklayer_export}
# FIRST a file list:
consensus_file_list <- list.files("analysis/00_consensus_peaks", full.names = T, pattern = ".bed")

# lapply (for loop) across consensus file list to add colnames
peaks <- lapply(consensus_file_list, read.table, col.names = c("chr", "start", "end", "name", "score", "strand"))
names(peaks) <- dbps
# double check order by looking at consensus_file_list is same order as dbps

# now clean up the chromosomes:
canonical_chr <- c(paste0("chr", 1:22), "chrM", "chrX", "chrY")
# let's use lapply with filter funciton to cannonical_chr
peaks <- lapply(peaks, function(x) x %>% filter(chr %in% canonical_chr))


# now that these are all nice and clean let's export:
new_filenames <- paste0("analysis/00_consensus_peaks/", names(peaks), "_consensus.bed")
for(i in 1:length(peaks)) {
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}


```

# Demo trying to load one of these peak files into UCSC.

# STEP4_3 create header on each file
Next we need a header to be compatible with UCSC 
We are using a broadPeak file which UCSC recognizes
but we need to tell it which one it is in header

There is a lot of infomation here on file formats for UCSC:

https://genome.ucsc.edu/FAQ/FAQformat.html#format1

The header we need is : track type=broadPeak name="
```{R adding header to peak files}

# we are using paste0 to print the header text
# then we are adding the names as the value the header equals with 'names' function

headers <- paste0("track type=bedGraph name=", names(peaks))
headers

# creating a path to export after we add header in for loop below
new_filenames <- paste0("analysis/00_consensus_peaks/ucsc_consensus_peaks/", names(peaks), ".bed")
new_filenames
```

Cool, now we have an object "new_filenames" that has 
the file path we want to write out each file to.
Let's use a for loop to print each file out separately and append the header.
```{R}
# let's do so more indexing and only print out the first two of the three files.

for(i in 1:length(peaks)) {
  # Write the header line
  writeLines(headers[[i]], new_filenames[[i]])
  # Append the broadPeak table data
  
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}

# note append=TRUE is very important otherwise when the peak data is imported it will overwrite header
```

# let's take a look in UCSC !
(1) file transfer a consensus peak to your comp
(2) go to class UCSC account BCHM_5631
(3) custome tracks and load consensus file.
(4) load raw data for your protein 
(5) compare consensus peaks with raw data!



###########################
ExeRcise: create function
###########################

Create a function and add to my_class_functions.R 
to format peak files for ucsc.



Solution


```{r exeRcise solution}



basepath <- "/scratch/Shares/rinnclass/CLASS_2023/<your_folder>"
consensus_peak_path <- "CLASS_2023/CLASSES/05_R_analyses/analysis/00_consensus_peaks"

# input into function
consensusFilePath <- file.path(basepath, consensus_peak_path)

# input into function
exportpath <- "analysis/00_consensus_peaks/test"






ucsc_formating <- function(consensusFilePath = consensusFilePath, export_path = export_path) {
  
consensus_file_list <- list.files(consensusFilePath, full.names = T, pattern = ".bed")

dbps <- sapply(consensus_file_list, function(x) {
y <- str_extract(x, "([^\\/]+$)")
unlist(strsplit(y, "_"))[1]})

peaks <- lapply(consensus_file_list, read.table, col.names = c("chr", "start", "end", "name", "score", "strand"))
names(peaks) <- dbps
print(length(peaks))
canonical_chr <- c(paste0("chr", 1:22), "chrM", "chrX", "chrY")
peaks <- lapply(peaks, function(x) x %>% filter(chr %in% canonical_chr))

headers <- paste0("track type=bedGraph name=", names(peaks))
new_filenames <- paste0("analysis/00_consensus_peaks/test/", names(peaks), ".bed")

for(i in 1:length(peaks)) {
  # Write the header line
  writeLines(headers[[i]], new_filenames[[i]])
  # Append the broadPeak table data
  
  write.table(peaks[[i]], new_filenames[[i]],
              sep = "\t", col.names = FALSE, row.names = FALSE,
              quote = FALSE, append = TRUE)
}
  
  return(c("done?"))
}


ucsc_formating(consensusFilePath = consensusFilePath, export_path = export_path)
  


```


















