---
title: "03_creating_sample_sheet"
author: "JR"
date: "7/22/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(httr)
library(janitor)
library(purrr)
source("/scratch/Shares/rinnclass/CLASS_2023/JR/CLASS_2023/util/class_functions.R")
source("/scratch/Shares/rinnclass/CLASS_2023/JR/CLASS_2023/util/encode_functions.R")
```
Goal: to take our sample sheet that we created with the encode API and 
get it ready for the NF_CORE chipseq pipeline

# cleaning up sample sheet

```{r sample sheet curration: adding replicate number for each RNAseq experiment}

# first let's read in our sample sheet from last class
samples <- read_csv("/scratch/Shares/rinnclass/CLASS_2023/JR/CLASS_2023/CLASSES/03_Nextflow/01_my_chipseq/samples.csv")

# For this experiment, the metadata in ENCODE was wrong.
# It seems to be single-end
# We can remove the paired-end identifier
samples[samples$experiment_accession == "ENCSR362CPB", "paired_end_identifier"] <- NA

# Let's number our replicates and create a new column called sample id 
# where we join the experiment accesion and the rep number
samples <- samples %>%
  # group_by clumps all the files together for a given exp accession
  group_by(experiment_accession) %>%
  # mutate to make new col rep_number using replicate col
  mutate(rep_number = as.numeric(factor(replicate))) %>%
  # combining together with unite
  unite(sample_id, experiment_accession, rep_number, sep = "_rep", remove = F)

# let's take a look
samples

```

# getting rid of non essential data
Now let's get rid of all that data we don't need! 

```{r removing data from sample sheet}
# We're just selecting a subset of columns from samples with dplyr::select
samples <- samples %>%
   dplyr::select(accession, sample_id, Target.gene.symbol, Controls, experiment_accession, Assay.title,
                Biosample.summary, md5sum, paired_end_identifier, rep_number) 


# Let's clean up the controls
samples <- samples %>%
  rowwise() %>%
  mutate(Controls = gsub("/experiments/|/", "", Controls),
         Controls = unlist(strsplit(Controls, ","))[[1]])
# now there is an issue that some are single end and paired end
# so let's make two data frames one for se and one for pe

se <- samples %>%
  filter(is.na(paired_end_identifier)) %>%
  mutate(fastq_file = accession) %>%

# note mutate luckily put this in the right spot to bind rows 
# later -- a solution would 
dplyr::select(accession, fastq_file, everything())

# now for paired end

pe <- samples %>%
  filter(!(is.na(paired_end_identifier))) %>%
  mutate(fastq_file = accession) %>%

# note mutate luckily put this in the right spot to bind rows 
# later -- a solution would 
dplyr::select(accession, fastq_file, everything())
# Now we have a new col fastq_file with the information for each fastq file
# We will keep adding to this until we get to the final NF_Core input file name.

samples <- bind_rows(pe, se)
# TODO: don't need to separate these here since we're not naming the file differently.
```

# Setting up filename for NF_CORE rnaseq pipeline
Now let's make the full filename for the fastq files. 
For the nf-core/rnaseq pipeline, the paired-end reads need to be named 
with the read number in the filename. 

```{r filename for NF_CORE rnaseq pipeline: adding .fastq.gz to fastq_file}
# We are now making a new column with .fastq.gz extension and then unite to make fastq file name.
samples <- samples %>%
  mutate(fq_extension = ".fastq.gz") %>%
  unite(fastq_file, fastq_file, fq_extension, sep = "", remove = F)
# nice we see a fastq file column and can be used as input into NF_CORE
```

# renaming fastq files to fit sample sheet
This cleaned up version of the samplesheet is good to go!
Now we want to rename the fastq files to the fastq name we just made.

```{r rename fastq files to samplesheet id}
# Rename the fastq files so that they contain the sample ID.

# The goal is to create a shell script that each row is a bash command.
 # TODO: this can be deleted.
# rename_script <- samples %>%
#   # removes grouping in samples
#   ungroup() %>%
#   dplyr::select(fastq_file, original_file) %>%
#   mutate(command = "mv") %>%
#   unite(command, command, original_file, fastq_file, sep = " ")
# 
# 
# # We can write this out as a bash script with ?write_lines
# ?write_lines
# 
# # We include a shebang header line so that the script is interpreted by bash.
# write_lines(c("#!/bin/bash", rename_script$command), "data/rename.sh")
```

# Now we will execute the shell script to mv file names

```{r bash shell script to change fastq file names}
# Now cd fastq and "chmod u+x rename.sh
# then ./rename.sh

# >>> voila all the files are renamed
```


# finalizing sample sheet for NF_CORE
Now we are going to take the steps to create a nicely formatted sample sheet 

```{r Reorganizing to make a sample sheet for DEseq2}
#### NOTE: don't need for CHIPSEq
# # Let's create the sample sheet that we will use later
# # to do the RNA-seq analysis in R.
# samples <- samples %>%
#   dplyr::rename(fastq = fastq_file,
#                 seq_type = Assay.title,
#                 sample_name = Biosample.summary) %>%
#   # The minus sign will remove this column -- which we no longer need.
#   dplyr::select(-original_file) 
```

# final organization of sample sheet

Now that we have it cleaned up, let's create one line for each replicate
where the fastq read 1 and read 2 are in the same row.

```{R organizing samplesheet }

# For this we will use the pivot wider function
# We need to tell the pivot_wider function which unique column combinations will specify each new row. 
# PIVOT WIDER by "paired_end_identifier" 
?pivot_wider

# first we have to do pivot wider for paired end
# single end isn't an issue since there is only one file
# group, replicate, fastq_1, fastq_2, antibody, control
pe_samplesheet <- samples %>%
  filter(!(is.na(paired_end_identifier))) %>%
# goal is to get read_1 and read_2 on the same line
  pivot_wider(id_cols = c("sample_id", "Target.gene.symbol", "Controls", "experiment_accession", "rep_number"),
              names_from = paired_end_identifier,
              values_from = c("fastq_file", "md5sum")) %>%
  group_by(Target.gene.symbol) %>%
  mutate(experiment_number = as.numeric(as.factor(experiment_accession))) %>%
  ungroup()
# Remove the controls that are not used
utilized_controls <- pe_samplesheet %>%
  filter(!is.na(Controls)) %>%
  pull(Controls) %>%
  unique()

pe_controls <- pe_samplesheet %>%
  filter(is.na(Controls),
         experiment_accession %in% utilized_controls)  %>%
  mutate(group = experiment_accession,
         antibody = "", control = "") %>%
  dplyr::rename(replicate = rep_number,
                fastq_1 = fastq_file_1,
                fastq_2 = fastq_file_2) %>%
  dplyr::select(group, replicate, fastq_1, fastq_2, antibody, control)

pe_chip <- pe_samplesheet %>%
  filter(!is.na(Controls)) %>%
  mutate(group = paste(Target.gene.symbol, experiment_number, sep = "_"),
         antibody = Target.gene.symbol) %>%
    dplyr::rename(replicate = rep_number,
                fastq_1 = fastq_file_1,
                fastq_2 = fastq_file_2,
                control = Controls) %>%
  dplyr::select(group, replicate, fastq_1, fastq_2, antibody, control)

pe_design <- bind_rows(pe_chip, pe_controls)
pe_design <- pe_design %>%
  arrange(control, group, replicate)
write_csv(pe_design, "results/pe_design.csv")


se_samplesheet <- samples %>%
  filter(is.na(paired_end_identifier)) %>%
  group_by(Target.gene.symbol) %>%
  mutate(experiment_number = as.numeric(as.factor(experiment_accession))) %>%
  ungroup()

utilized_controls <- se_samplesheet %>%
  filter(!is.na(Controls)) %>%
  pull(Controls) %>%
  unique()

se_controls <- se_samplesheet %>%
  filter(is.na(Controls),
         experiment_accession %in% utilized_controls)  %>%
  mutate(group = experiment_accession,
         antibody = "", control = "",
         fastq_2 = "") %>%
  dplyr::rename(replicate = rep_number,
                fastq_1 = fastq_file) %>%
  dplyr::select(group, replicate, fastq_1, fastq_2, antibody, control)

se_chip <- se_samplesheet %>%
  filter(!is.na(Controls)) %>%
  rowwise() %>%
  mutate(group = paste(Target.gene.symbol, experiment_number, sep = "_"),
         antibody = Target.gene.symbol,
         fastq_2 = "") %>%
    dplyr::rename(replicate = rep_number,
                fastq_1 = fastq_file,
                control = Controls) %>%
  dplyr::select(group, replicate, fastq_1, fastq_2, antibody, control)
se_design <- bind_rows(se_chip, se_controls)
# Make sure it's in the proper order
se_design <- se_design %>%
  arrange(control, group, replicate)

write_csv(se_design, "results/se_design.csv")
```

# After index is saved add this flag and take away save ref flag
 --bwa_index
type: 'string'
Full path to directory containing BWA index including base name. i.e. /path/to/index/genome.fa.

# first run add this to get bwa index for a given genome

 --save_reference
type: 'boolean'
If generated by the pipeline save the BWA index in the results directory.

Save in genomes














