---
title: "01_R_studio_install_packages_basics"
author: "JR"
date: "7/20/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(tidyverse)
library(GenomicRanges)
```

Today we are going to connect to Rstudio on the Fiji-viz server.

Accompanying Youtube Video:
https://youtu.be/2ESg9rpGQr8


We simply go here and login with our identikey:
https://fiji-viz.colorado.edu/
(click Rstudio button) -> login

# If you are taking this course elsewhere you can run most of the code on a local Rstudio (free)
https://www.rstudio.com/products/rstudio/download/

Please see video for an orientation of Rstudio:
https://www.youtube.com/watch?v=5SnLzPKa8bM

# In class we will login and make our new R project for class.
Please make sure you clone the class repo from git here:
/scratch/Shares/rinnclass/CLASS_2023/<Your folder>
# Pleae see this video for how to clone and connect github:
https://youtu.be/dTlz2sJgv1A


#################################
# 1 Make R project in Rstudio
#################################

Let's start by making an Rpoject -- that will also connect to git.
In the top right hand corner there is a dropdown with "R shield"

(a) select open new project
(b) select from an existing folder (the class_2023 folder you cloned or forked)
(c) name the project.

#################################
# 2 Install packages
#################################

Now we will use Install.packages() for most packages. This will install any packages in CRAN repository. Let's try one.
```{R install packages}
?install.packages
install.packages("tidyverse")
```
Cool we have our first package installed.

However sometimes we need to install from Bioconductor and need to use BiocManager::install function
```{R install BiocManager}
if (!requireNamespace("BiocManager", quietly = TRUE))
install.packages("BiocManager", version = '3.10')
```

Let's now install genomic ranges package we will be using for most of class
# note we will have two classes on "Iranges" and Granges before we use GenomicRanges package
```{R install genomic ranges}
# Here is an example BiocManager::install
BiocManager::install("GenomicRanges")
# try installing all the packages listed below.
# note use either install.packages or BiocManager::install
```


################
# 3 {r setup, }
################
Note at the top of the document we source packages and functions we created.
The whole chunk looks like this:
```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
# this is providing information on how to "knit" the .RMD file.
# You can make really pretty results and corresponding code.
# Here is an example of a knitted document from last year:
# /scratch/Shares/rinnclass/CLASS_2022/EricY/bchm/exercises/analysis/final

options(stringsAsFactors = F)
# This is an important thing to have. When you read in a table R default is to
# make the columns have a "factor" associated with them. This is like a "meta variable"
# The col still has a name but that name is now an associated factor above the name.
# When we run RNAseq pipelines the comparison samples are "factorized" so the program
# knows which cols represent which samples for differential expression etc.

library(tidyverse)
library(GenomicRanges)
# This is how we load packages with "library(package X)

```


############################################
Exercise : install packages needed for class
############################################

# Hint google the package name and see if it is CRAN (install.packages)
# Or if in Bioconductor (BiocManager::install)

# probably worth reading what they do while installing...

Iranges
ggplot2
broom
Gviz
ggdendro
ggrepel
pheatmap
httr
janitor
purrr
deseq2

# let's try this one -- maybe tricky:
ggpubr








